Ref/SignalGen/SignalGen.cpp

```diff
@@ -38,153 +38,222 @@ namespace Ref {
+        this->signalPhase = Phase;
+        this->sigType = SigType;
 
+        // When the settings change, reset the history values
+        for (U32 i = 0; i < SignalSet::SIZE; i++) {
+            this->sigHistory[i] = 0.0f;
+        }
+        for (U32 i = 0; i < SignalPairSet::SIZE; i++) {
+            this->sigPairHistory[i].settime(0.0f);
+            this->sigPairHistory[i].setvalue(0.0f);
+        }
+        this->log_ACTIVITY_LO_SignalGen_SettingsChanged(this->signalFrequency, this->signalAmplitude, this->signalPhase, this->sigType);
+        this->tlmWrite_Type(SigType);
+        this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::OK);
+    }
 
+    void SignalGen::SignalGen_Toggle_cmdHandler(
+        FwOpcodeType opCode, /*!< The opcode*/
+        U32 cmdSeq /*!< The command sequence number*/
+    )
+    {
+        this->running = !this->running;
+        this->ticks = 0;
+        this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::OK);
+    }
 
-  void SignalGen :: SignalGen_Skip_cmdHandler(
-      FwOpcodeType opCode, /*!< The opcode*/
-      U32 cmdSeq /*!< The command sequence number*/
+    void SignalGen::SignalGen_Skip_cmdHandler(
+        FwOpcodeType opCode, /*!< The opcode*/
+        U32 cmdSeq /*!< The command sequence number*/
     )
-  {
-      this->skipOne = true;
-      this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::OK);
-  }
+    {
+        this->skipOne = true;
+        this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::OK);
+    }
+
+    void SignalGen ::
+        SignalGen_Dp_cmdHandler(
+            FwOpcodeType opCode,
+            U32 cmdSeq,
+            U32 records
+        )
+    {
+        // make sure DPs are available
+        if (
+            not this->isConnected_productGetOut_OutputPort(0)
+        ) {
+            this->log_WARNING_HI_SignalGen_DpsNotConnected();
+            this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::EXECUTION_ERROR);
+            return;
+        }
+
+        // get DP buffer
+        this->dpGet_DataContainer(records*(SignalInfo::SERIALIZED_SIZE + sizeof(FwDpIdType)),this->m_dpContainer);
+        this->m_dpInProgress = true;
+        this->m_numDps = records;
+        this->m_currDp = 0;
+        this->log_ACTIVITY_LO_SignalGen_DpStarted(records);
+        this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::OK);
+    }
+
+    void SignalGen::cleanupAndSendDp() {
+        this->dpSend(this->m_dpContainer);
+        this->m_dpInProgress = 0;
+        this->m_dpBytes = 0;
+        this->m_numDps = 0;
+        this->m_currDp = 0;
+    }
+
+
+
+    // ----------------------------------------------------------------------
+    // Handler implementations for data products
+    // ----------------------------------------------------------------------
+
+    void SignalGen ::
+        dpRecv_DataContainer_handler(
+            DpContainer& container,
+            Fw::Success::T status
+        )
+    {
+        // TODO
```

Comment:
This should be removed or demonstrated.

Reply:
I was going to add that update to `SignalGen` after the demo was merged.

Reply:
@timcanham this was merged yesterday.

Reply:
Ah... Okay, I'll do another PR for the update.